axum-macros/src/debug_handler/specializer.rs

```diff
@@ -0,0 +1,199 @@
+use std::collections::{HashMap, HashSet};
+
+use itertools::Itertools;
+use proc_macro2::TokenStream;
+use syn::{
+    visit::{self, Visit},
+    visit_mut::{self, VisitMut},
+    GenericParam, ItemFn,
+};
+
+use quote::quote;
+
+use super::attr::SpecializationsAttr;
+
+struct GenericFinder<'a> {
+    found_idents: HashSet<&'a syn::Ident>,
+    generic_param_set: HashSet<&'a syn::Ident>,
+}
+
+impl<'ast, 'a> Visit<'ast> for GenericFinder<'a> {
+    fn visit_ident(&mut self, ident: &'ast syn::Ident) {
+        if let Some(a) = self.generic_param_set.get(ident) {
+            self.found_idents.insert(a);
+        }
+        // Delegate to the default impl to visit nested expressions.
+        visit::visit_ident(self, ident);
+    }
+}
+
+struct TypeSpecializer<'a> {
+    specializations: &'a HashMap<&'a syn::Ident, &'a syn::Type>,
+}
+
+impl<'a> VisitMut for TypeSpecializer<'a> {
+    fn visit_type_mut(&mut self, ty: &mut syn::Type) {
+        match ty {
+            syn::Type::Path(ty_path) => {
+                if ty_path.path.segments.len() == 1 {
+                    let ident = &ty_path.path.segments[0].ident;
+                    if let Some(specialized) = self.specializations.get(ident) {
+                        *ty = (*specialized).clone();
+                        return; // don't recuresively visit substituted values
+                    }
+                }
+            }
+            _ => (),
+        };
+        visit_mut::visit_type_mut(self, ty);
+    }
+}
+pub(crate) struct Specializer {
+    generic_params: Vec<syn::Ident>,
+    specializations: HashMap<syn::Ident, Vec<syn::Type>>,
+}
+
+impl Specializer {
+    pub(crate) fn new(
+        with_tys: Option<SpecializationsAttr>,
+        item_fn: &ItemFn,
+    ) -> Result<Self, syn::Error> {
+        let specializations = with_tys.map(|f| f.specializations).unwrap_or_default();
+
+        let generic_params = item_fn
+            .sig
+            .generics
+            .params
+            .iter()
+            .enumerate()
+            .map(|(_idx, param)| match param {
+                GenericParam::Type(t) => Ok(t.ident.clone()),
+                _ => Err(syn::Error::new_spanned(
+                    param,
+                    "Only type params are supported by `#[axum_macros::debug_handler]`.",
+                )),
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        for param in &generic_params {
+            if !specializations.contains_key(param) {
+                return Err(
+                    syn::Error::new_spanned(
+                        param,
+                        "Generic param is missing a specialization in `#[axum_macros::debug_handler]`. Specify each generic param at least once using the 'with' attribute to support debugging generic functions.",
+                    )
+                );
+            }
```

Comment:
Can't this check be moved into the first arm of `match param` inside the `map` closure?